### PR TITLE
CPD: add support for Typescript using antlr4 grammar

### PR DESCRIPTION
**Pull request by pguyot**
## Describe the PR

Following T-SQL, add minimal CPD support for Typescript using antlr4 grammar from https://github.com/antlr/grammars-v4

## Related issues

- Fixes #

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pguyot** [comment]
TypescriptLexerBase, coming directly from antlr4 repository, doesn't pass checkstyle.
Please advise what would be the best approach.

File: pmd-typescript/src/main/java/net/sourceforge/pmd/lang/typescript/ast/TypeScriptLexerBase.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+// Source: https://github.com/antlr/grammarsv4/
+
+import org.antlr.v4.runtime.*;
+
+import java.util.Stack;
+
+/**
+ * All lexer methods that used in grammar (IsStrictMode)
+ * should start with Upper Case Char similar to Lexer rules.
```
Comment (oowekyala) [inline pmd-typescript/src/main/java/net/sourceforge/pmd/lang/typescript/ast/TypeScriptLexerBase.java:11]: > TypescriptLexerBase, coming directly from antlr4 repository, doesn't pass checkstyle. Please advise what would be the best approach.

Can you try renaming the methods to use regular camel case? AFAIK this casing convention is not necessary but just an internal convention the authors used in their codebase. Also just format the file, there are some obvious style inconsistencies that checkstyle rightfully reports on (brace placement eg).

File: pmd-typescript/src/main/java/net/sourceforge/pmd/lang/typescript/ast/TypeScriptLexerBase.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ */
+public abstract class TypeScriptLexerBase extends Lexer
```
Comment (oowekyala) [inline pmd-typescript/src/main/java/net/sourceforge/pmd/lang/typescript/ast/TypeScriptLexerBase.java:13]:
```suggestion
abstract class TypeScriptLexerBase extends Lexer
```

**Review by oowekyala — COMMENTED** [review]
Thanks for the contribution!

I think it would be nice to put this all into the existing `pmd-javascript` module though.

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No regression tested rules have been changed.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-javascript/src/main/java/net/sourceforge/pmd/lang/typescript/cpd/TypescriptLanguage.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import net.sourceforge.pmd.cpd.AbstractLanguage;
+
+/**
+ * @author pguyot@kallisys.net
+ */
+public class TypescriptLanguage extends AbstractLanguage {
+
+    public TypescriptLanguage() {
+        super("Typescript", "typescript", new TypescriptTokenizer(), ".ts");
```
Comment (oowekyala) [inline pmd-javascript/src/main/java/net/sourceforge/pmd/lang/typescript/cpd/TypescriptLanguage.java:15]: I will rename this "TypeScript" with capital S

**Review by oowekyala — APPROVED** [review]
Thanks!

**Pull request merged**